src/overlays/actors/ovl_Bg_Dblue_Elevator/z_bg_dblue_elevator.c

```diff
@@ -5,68 +5,373 @@
+        var_fs0 = Rand_ZeroOne();
+        var_fs0 = 1.0f - (var_fs0 * var_fs0);
+        rn = Rand_Next();
+        if (rn > 0) {
+            var_fs0 = -var_fs0;
+        }
+        ripplePos.z = (var_fs0 * 100.0f) + this->dyna.actor.world.pos.z;
+        EffectSsGRipple_Spawn(play, &ripplePos, 400, 800, sSmallRipplesLives[i]);
+    }
+    Matrix_Pop();
+}
+
+s32 BgDblueElevator_GetWaterFlowFromCeiling(Actor* thisx, PlayState* play) {
+    BgDblueElevator* this = THIS;
+
+    if (Flags_GetSwitch(play, BG_DBLUE_ELEVATOR_GET_7F(&this->dyna.actor, 0))) {
+        return 0;
+    }
+    return 1;
+}
+
+s32 BgDblueElevator_GetWaterFlowFromPipes(Actor* thisx, PlayState* play) {
+    BgDblueElevator* this = THIS;
+    s32 var_s0 = 0;
+
+    // This checks if the first the yellow tank turnkey is open.
+    if (!Flags_GetSwitch(play, BG_DBLUE_ELEVATOR_GET_7F(&this->dyna.actor, 0))) {
+        var_s0 = 1;
+    }
+
+    // This checks if water is flowing from the red pipes.
+    if (Flags_GetSwitch(play, BG_DBLUE_ELEVATOR_GET_7F(&this->dyna.actor, 1)) &&
+        Flags_GetSwitch(play, BG_DBLUE_ELEVATOR_GET_7F(&this->dyna.actor, 2)) &&
+        Flags_GetSwitch(play, BG_DBLUE_ELEVATOR_GET_7F(&this->dyna.actor, 3))) {
+        var_s0 += 2;
+    }
+    return var_s0;
+}
+
+void BgDblueElevator_Init(Actor* thisx, PlayState* play2) {
+    BgDblueElevator* this = THIS;
+    PlayState* play = play2;
+    s32 index = BG_DBLUE_ELEVATOR_GET_INDEX(&this->dyna.actor);
+    BgDblueElevatorStruct1* ptr = &D_80B92960[index];
+    s32 waterFlow;
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyActor_Init(&this->dyna, DYNA_TRANSFORM_POS);
+    DynaPolyActor_LoadMesh(play, &this->dyna, &gGreatBayTempleObjectElevatorCol);
+    waterFlow = ptr->getWaterFlow(this, play);
+
+    if (waterFlow == BG_DBLUE_ELEVATOR_WATER_FLOW_REVERSED) {
+        this->direction = -ptr->initialDirection;
+    } else {
+        this->direction = ptr->initialDirection;
+    }
+
+    BgDblueElevator_CheckWaterBoxInfo(this, play);
+
+    if ((waterFlow == BG_DBLUE_ELEVATOR_WATER_FLOW_STOPPED) ||
+        (waterFlow == BG_DBLUE_ELEVATOR_WATER_FLOW_BOTH_DIRECTIONS)) {
+        BgDblueElevator_SetupWaitForMove(this);
+    } else {
+        BgDblueElevator_SetupMove(this);
+    }
+}
+
+void BgDblueElevator_Destroy(Actor* thisx, PlayState* play) {
+    BgDblueElevator* this = THIS;
+
+    DynaPoly_DeleteBgActor(play, &play->colCtx.dyna, this->dyna.bgId);
+}
+
+void BgDblueElevator_SetupWaitForMove(BgDblueElevator* this) {
+    this->activationTimer = 60;
+    this->actionFunc = BgDblueElevator_WaitForMove;
+}
+
+/**
+ * The elevator waits for 3 seconds (60 frames) when water restarts flowing before resuming its halted cycle.
+ */
+void BgDblueElevator_WaitForMove(BgDblueElevator* this, PlayState* play) {
+    s32 waterFlow;
+    s32 index = BG_DBLUE_ELEVATOR_GET_INDEX(&this->dyna.actor);
+
+    waterFlow = (&D_80B92960[index])->getWaterFlow(this, play);
```

Comment:
```suggestion
    s32 index = BG_DBLUE_ELEVATOR_GET_INDEX(&this->dyna.actor);
    BgDblueElevatorStruct1* ptr = &D_80B92960[index];
    s32 waterFlow = ptr->getWaterFlow(this, play);

```
Does this work? In general this pattern seems pretty prevalent in a lot of the functions below as well, and personally I think I like the extra `ptr` variable in each case if possible.